Adds initial documentation

trozet opened the pull request:
Sets up the specs repo with basic information along with some of the
initial architecture documentation.

trozet committed:
Adds initial documentation

Sets up the specs repo with basic information along with some of the
initial architecture documentation.

shague commented on README.md:
```diff
@@ -1,2 +1,29 @@
+projects.
+
+## Documentation Overview
+The core folder contains all of the fundamental core design and developer
```
core and fundamental are redundant

trozet replied:
Got it will just use fundamental

shague commented on README.md:
```diff
@@ -1,2 +1,29 @@
-Specs and design documentation
+# Nimbess - Specs
+Nimbess is a distributed controller architecture designed to unify network
+policy across Kubernetes. The docs repo contains documentation about Nimbess
```
policy or intent?

trozet replied:
Should we say something like this?
"Nimbess is a distributed controller architecture designed to unify network policy, as well as express network intent across Kubernetes."

shague commented on README.md:
```diff
@@ -1,2 +1,29 @@
-Specs and design documentation
+# Nimbess - Specs
+Nimbess is a distributed controller architecture designed to unify network
+policy across Kubernetes. The docs repo contains documentation about Nimbess
```
Kubernetes ot OpenShift

trozet replied:
Kubernetes, lets use upstream terms

shague commented on README.md:
```diff
@@ -1,2 +1,29 @@
-Specs and design documentation
+# Nimbess - Specs
+Nimbess is a distributed controller architecture designed to unify network
+policy across Kubernetes. The docs repo contains documentation about Nimbess
```
Nimbess'

trozet replied:
Done

trozet commented on core/architecture.md:
```diff
@@ -0,0 +1,82 @@
+container. Technically speaking, a port just points to a socket referencing
+virtual or physical memory address space. It does not matter if there is 1 or
+more ports between a a dataplane and a container. Functionally, it matters for
+traffic differentiation between a pod control plane, and perhaps one more more
```
"and perhaps one or more"

trozet commented on core/architecture.md:
```diff
@@ -0,0 +1,82 @@
+the network and then configures it appropriately. This key point means that
+rather than defining network configuration within pod annotations at pod
+instantiation or deletion, pod's networking configuration is done separately as
+a UNP resource declaration later. This means that the pods network
```
pod's

shague commented on README.md:
```diff
@@ -1,2 +1,29 @@
-# docs
-Specs and design documentation
+# Nimbess - Specs
```
Does Nimbess always require the dataplane to be BESS? Or do we want something more generic? If always Bess, then NimBESS.

trozet replied:
Not always bess, data plane will be abstracted into several drivers. BESS is default.

trozet commented on core/architecture.md:
```diff
@@ -0,0 +1,82 @@
+In container networking, there is no real reason to use more than one port to a
+container. Technically speaking, a port just points to a socket referencing
+virtual or physical memory address space. It does not matter if there is 1 or
+more ports between a a dataplane and a container. Functionally, it matters for
```
remove duplicate "a"

shague commented on core/architecture.md:
```diff
@@ -0,0 +1,82 @@
+## Overview
+
+Nimbess is a light-weight distributed controller architecture that runs as a
+Kubernetes pod across every Kubelet node. Nimbess by default supports the
```
Don't think you need Kubelet, as a Node is for container workloads.

fepan replied:
We don't need detail here about how the controller is run. I think it's more important to describe it a micro-service architecture that is completely distributed and can run in any cloud environment.

trozet replied:
Will re-word this

shague commented on core/architecture.md:
```diff
@@ -0,0 +1,82 @@
+Nimbess is a light-weight distributed controller architecture that runs as a
+Kubernetes pod across every Kubelet node. Nimbess by default supports the
+Berkley Extensible Software Switch [BESS](https://github.com/NetSys/bess)
+dataplane in order to achieve maximum network pipeline programmability and
```
Today we think BESS has the best performance, but can we truly say that it is? Or should we have real details why we think BESS is better?

shague commented on README.md:
```diff
@@ -1,2 +1,29 @@
-# docs
-Specs and design documentation
+# Nimbess - Specs
+Nimbess is a distributed controller architecture designed to unify network
```
what does  unify network policy really mean? What is being unified?

trozet replied:
unifying network control into a single policy CRD, rather than specifying it in different resources throughout k8s

shague reviewed (commented):
Just a start to comments.

fepan commented on core/architecture.md:
```diff
@@ -0,0 +1,82 @@
+and refers only to access-control policy of "who can talk to who". 
+
+One of the main goals of Nimbess is to simplify and consolidate these features
+into a single point of entry using a custom resource definition (CRD) called
```
I think our stated main goal should not be to consolidate features, but to provide an entirely new way of thinking about container or kubernetes networking that's based on policy and intent and defining a mechanism to express and realize them.

trozet replied:
yes, will reword it

fepan commented on core/architecture.md:
```diff
@@ -0,0 +1,82 @@
+CNIs, and thus multiple ports to be created.
+
+Nimbess resolves both of these issues in different ways. By default, when a pod
+is created, Nimbess places a single port for that pod in a default network. It
```
change "in a default network" to "with default connectivity"

trozet replied:
yeah and need to update this to our 2 port (1 kernel, 1 dpdk) current proposal (may change again in the future)

trozet commented on core/architecture.md:
```diff
@@ -0,0 +1,82 @@
+Service Function Chaining (SFC), etc.
+
+Additionally, CNI implementations are also inherently limited in Kubernetes,
+due the nature of the specification which limits CNI invocation to pod create
```
due to the nature of

shague commented on README.md:
```diff
@@ -1,2 +1,29 @@
+## Feature Specs
+Features should be added with a new folder within the features folder, along
+with a spec written detailing the changes and impact of the new feature. Follow
+the [example](features/example/example.md) template when adding a new feature.
```
wonder if the example.md should be name example_spec.md?

shague commented on README.md:
```diff
@@ -1,2 +1,29 @@
+* [Developer Guide](core/developer.md)
+
+## Feature Specs
+Features should be added with a new folder within the features folder, along
```
Sentence seems a little long...
All new features should include a Feature Spec that details the changes and impacts of the new feature. Create a folder for each new feature.

trozet replied:
done

shague commented on README.md:
```diff
@@ -1,2 +1,29 @@
+## Feature Specs
+Features should be added with a new folder within the features folder, along
+with a spec written detailing the changes and impact of the new feature. Follow
+the [example](features/example/example.md) template when adding a new feature.
```
Should we have a specific spec template at the root and then an example?

trozet replied:
I don't think it's necessary, but I'm not against it. Can add it in a follow up patch if you want.

shague commented on README.md:
```diff
@@ -1,2 +1,29 @@
+request to this repo.
+
+## License
+Nimbess is available under hte Apache 2.0 license. See [LICENSE](LICENSE) file
```
s/hte/THe

trozet replied:
done

shague commented on core/architecture.md:
```diff
@@ -0,0 +1,150 @@
+## Overview
+
+Nimbess is a light-weight, micro-service controller architecture that is
+completely distributed and can be run in any cloud environment. Nimbess by
```
Is this accurate - can be run an any cloud - or is that a goal?

trozet replied:
yeah we should be able to run on any cloud because our software will be container native, and use overlay to overcome different networking envs

shague commented on core/architecture.md:
```diff
@@ -0,0 +1,150 @@
+minimum required bandwidth of 25 Gbps. The policy would be translated on the
+backend in order to recognize that a high speed path is required, and
+reconfigure QoS and ACL in order to apply the requested intent across the
+cluster’s network. UNP is not only limited to  access-control lists (ACL),
```
remove extra space

trozet replied:
done

shague commented on core/architecture.md:
```diff
@@ -0,0 +1,150 @@
+for high speed interfaces with CNFs, there is also a need to supply a kernel
+interface for default Kubernetes networking and CNF management. With Nimbess,
+a single kernel port for default network connectivity is provided, while also
+providing a high-speed secondary interface. Nimbess then relies on a UNP in
```
Should we just say secondary interface and leave it to the user to decide what it is for? We should list an example of why the second interface is needed, like if they want high-speed. It seems better to describe it as the primary k8s port and then the secondary port.

trozet replied:
The secondary interface can be used however a user wants, but it is still a high-speed interface.

shague commented on core/architecture.md:
```diff
@@ -0,0 +1,150 @@
+configures the secondary interface appropriately. This key point means that
+rather than defining network configuration within pod annotations at pod
+instantiation or deletion, pod's networking configuration is done separately as
+a UNP resource declaration. This means that the pod's network configuration is
```
Does any port config need to be included at pod creation? Will the creation fail if nothing is included in the port config?

trozet replied:
Typical required CNI configuration per the CNI spec lists what fields are required to be present for an ADD request. We follow the spec so yes, certain fields are required.

shague commented on core/architecture.md:
```diff
@@ -0,0 +1,150 @@
+rather than defining network configuration within pod annotations at pod
+instantiation or deletion, pod's networking configuration is done separately as
+a UNP resource declaration. This means that the pod's network configuration is
+updatable at runtime by simply updating the policy. This different from NSM
```
This IS different

trozet replied:
Done

shague commented on core/architecture.md:
```diff
@@ -0,0 +1,150 @@
+updatable at runtime by simply updating the policy. This different from NSM
+for three reasons:
+
+1. Nimbess only ever uses a single high-speed port for a pod
```
This is confusing - we made a decision to only support a single secondary port, but how does that conflict with NSM? NSM doesn't have to have more ports though it can handle it. Or maybe adding details about NSM here would help to highlight the contrasts. List these bullets here as simple Nimbess architectural points and then list the contrasts to NSM.

trozet replied:
NSM always adds more ports for multiple connections from a pod. I'll add that to the bullet point.

shague commented on core/architecture.md:
```diff
@@ -0,0 +1,150 @@
+traffic differentiation between a pod control plane, and perhaps one or more
+data plane networks it is connected to. However, the goal of Nimbess is to 
+simplify networking, and one way to do that is to do real traffic
+differentiation within the data plane itself. Therefore with UNP, one can set
```
Use wording like "As an example"?

trozet replied:
Done

shague commented on core/architecture.md:
```diff
@@ -0,0 +1,150 @@
+2. UNP does not try to do VNF/CNF orchestration with K8S
+3. UNP is not restricted to L2-L3 service connectivity/policy
+
+In container networking, there is no real reason to use more than one port to a
```
hmm, maybe my previous comment isn't so important since the description of Nimbess is here. Guess it is flipped from what I was thinking. Leave up to you if you want to reword it.

shague commented on core/architecture.md:
```diff
@@ -0,0 +1,150 @@
+
+In container networking, there is no real reason to use more than one port to a
+container. Technically speaking, a port just points to a socket referencing
+virtual or physical memory address space. It does not matter if there is 1 or
```
Did we talk about what it means if k8s is adding policy for the primary port and UNP is used for the secondary when it comes to the data plane? Meaning the data plane will also implement the policy for the primary port, but can that conflict the UNP? ANd do we intend to block k8s policy on the secondary ports and allow only UNP?

trozet replied:
In my opinion, if any POD has a UNP applied to it, that should take precedence over any other k8s overlapping resource.

shague commented on core/architecture.md:
```diff
@@ -0,0 +1,150 @@
+
+### Nimbess Stargazer
+Stargazer is responsible for watching k8s resources and reflecting them into
+the Nimbess etcd. Stargazer also handles delegating work to Nimbess Agents if
```
I don't recall this delegation piece. I thought the agents only responded to etcd updates.

trozet replied:
I've had a lot of comments on this wording of this one sentence (delegating vs scheduling vs assigning, etc). The main point here is that somehow stargazer and the agents work together to handle translating k8s resources into Nimbess network modules. Some ideas include a worker queue built by Stargazer which then agent's subscribe to and each one takes a job. We don't have that piece totally ironed out yet, so lets not harp on this one sentence and we can address/fix this later once we know exactly how we are going to implement it.

shague commented on core/architecture.md:
```diff
@@ -0,0 +1,150 @@
+Calico is used to provide the BGP stack and route insertion into the kernel.
+
+### Climate Control
+Climate Control serves as the CLI tool for interacting directly with Nimbess in
```
It could do more that inspection right? Or is that what you meant bu Nimbess specific configuration? If so reword to "and set Nimbess specific configuration".

oglok replied:
In the drawing, Climate control is accessing directly the Nimbess Etcd cluster, instead of making calls to BESS itself. What's the right workflow?
Is this a pre-existing tool? or something you're planning to develop for traceability and troubleshooting purposes?

trozet replied:
This is a totally new tool that honestly we know we need but have not ironed out exactly how this works yet. We can update the drawing and this document once we have a better idea of how we want to implement this. For now, the agent, CNI client, and stargazer are the most important pieces to focus on first.

shague commented on core/architecture.md:
```diff
@@ -0,0 +1,150 @@
+
+### Doppler
+Doppler provides a customized GUI in order to visually examine the k8s cluster
+network, as well as visualize the data plane network fabric.
```
Any option to modify configuration or some troubleshooting functions like initiating packet traces.

trozet replied:
That's a good idea. We have that capability to tap anywhere in any BESS pipeline. I'll reword this.

shague commented on features/example/example.md:
```diff
@@ -0,0 +1,25 @@
+# Example feature spec
+This contains an example of a new feature to be added.
```
Related to earlier comment... but this line here could be: <A short description of the feature>. And that is what would be in the template. But the example here is what you have now. Same for the sections below. Or it could be merged into this example.

shague commented on features/example/example.md:
```diff
@@ -0,0 +1,25 @@
+List any concerns around user experience with your change.
+
+## Owners
+List all owners that will collaborate to implement this feature.
```
Add a test section to ensure we have tests?

trozet replied:
How do you want that section to look? Contributor provides an overview of the test cases or areas that will need testing?

trozet replied:
I added a section requesting an overview of testing required.

shague commented on features/example/example.md:
```diff
@@ -0,0 +1,25 @@
+List any concerns around performance increase/decrease with your change.
+
+## User Impact
+List any concerns around user experience with your change.
```
Upgrade-ability or backward compatibility concerns?

trozet replied:
done

shague commented on features/example/example.md:
```diff
@@ -0,0 +1,25 @@
+## Performance Impact
+List any concerns around performance increase/decrease with your change.
+
+## User Impact
```
external impacts or concerns?

trozet replied:
Please elaborate on this. Not sure what you mean.

shague replied:
Do we want to add sections for external impacts like to other projects for example?

trozet replied:
OK will add that.

oglok commented on core/architecture.md:
```diff
@@ -0,0 +1,150 @@
+
+[![Nimbess Architecture](../img/Nimbess.png)](../img/Nimbess.png)
+
+### Nimbess ETCD
```
Does this mean that Nimbess declarative state will run on a separate etcd cluster than the K8s one? If so, what's the reason for doing that?

fepan replied:
Yes. The etcd in k8s is used for storing infra information. we should be deploying our own data store to avoid issues like collision, security (permission) and availability.

oglok replied:
I see Nimbess Stargaze as a K8s controller watching the UNP, and then, you could store the extra information needed for the implementation of the network intent in a separate etcd cluster specific for Nimbess. Copying the same info that is stored in K8s etcd to a separate one doesn't look very eficient to me. Just opening this for debate, and I know Calico and Contiv do that.

trozet replied:
Your point is well taken that we should not be duplicating objects in databases if we don't need to. Our thought is that we may need to update the resources with specific Nimbess fields like the ID of the resource that we translate it to. For example, if we see a k8s service, and we translate that into a load balancer Nimbess network module, then we may want to update that k8s service resource with the ID of our load balancer so we can back reference it later. We haven't worked through the specifics of how that part will work yet, but will include you when we come up with the design.

trozet committed:
Updates doc based on comments

Updates Nimbess picture and addresses missing section in example and
typos in architecture doc.

trozet committed:
Addresses external impacts

Adds external impact section to new feature reqs.

trozet committed:
Addresses shague latest comments

Fixes typos and addresses comment about port differences with NSM.
